### DISCUSSION TITLE
How much should `fc` bandwidth be over-estimated for SetGaussExcite?

### DISCUSSION THREAD

**Discussion by KJ7LNW** [thliebig/openEMS-Project#165] [Q&A Usage] [answered]
The 2-meter band that I am targeting for antenna design has a 4 MHz bandwidth available in my license class. 

Example:

```py
f0 = 146e6
fc = 2e6
...
FDTD.SetGaussExcite( f0, fc )
```

Questions:

- What happens to simulation accuracy as the frequency approaches the minimum of `f0-fc == 144e6`?  
- What does "-20dB cutoff"  mean in this context, as I have seen in the tutorial example comments?
- How much should `fc` be overestimated to ensure accurate simulation results in the bandwidth we wish to cover?

**Comment by VolkerMuehlhaus** [comment, marked as answer]
> What happens to simulation accuracy as the frequency approaches the minimum of f0-fc == 144e6?
> What does "-20dB cutoff" mean in this context, as I have seen in the tutorial example comments?

f0 defines the center frequency of the spectrum created by the exciation pulse, and fc defines the frequency offset where energy is 20dB down from the maximum (at f0). 

Rule of thumb: for S-parameter results in FDTD we want the frequencies of interest within the -20dB bandwidth. Note that fc is the offset for -20dB, so the total -20dB **bandwidth** around f0 is 2*fc

> How much should fc be overestimated to ensure accurate simulation results in the bandwidth we wish to cover?

In my antenna simulation, I use excitations that are much wider than the actual antenna bandwidth. We can always limit the plotted frequency range to values less than the excitation bandwidth. This is an FFT postprocessing step, so we can even re-run that part without repeating the FDTD simulation, to try different plot ranges. 

Having an excitation with large bandwidth means we get reflection across a wide frequency range, but that doesn't hurt us because the reflected signal is properly absorbed by the port.

One reason why I don't choose very narrow excitation bandwidth is the time response of the excitation signal: smaller relative bandwidth means a wider pulse package for the excitation signal, at a given simulation timestep. Below is an example, note the different time axis.

Of course, multi-path reflections inside the model will cause time signals at the ports that last longer than the initial excitation pulse, but my understanding and experience is that a very stretched excitation pulse (small fc/f0 ratio) will take more timesteps until the simulation is converged.

![excitation_timesignal](https://github.com/thliebig/openEMS-Project/assets/47922981/b706e25d-50ab-43e3-9adb-9c3cd73d313d)

**Reply by KJ7LNW** [reply]
> Of course, multi-path reflections inside the model will cause time signals at the ports that last longer than the initial excitation pulse, but my understanding and experience is that a very stretched excitation pulse (small fc/f0 ratio) will take more timesteps until the simulation is converged.

Good to know!  I had assumed the opposite, thinking more bandwidth meant more to simulate (ie, like in frequency-domain simulations).  I'm simulating for a 4MHz bandwidth at 146MHz (144-148MHz) so I'll choose something much wider.

- What problems might be associated with `fc` being too wide?
- What happens if `fc` is greater than `f0` (ie, `f0-fc < 0`)?

**Reply by VolkerMuehlhaus** [reply]
>  What happens if fc is greater than f0 (ie, f0-fc < 0)?

This is still valid. For example in the transmission line case below, f0=0 and fc=5 GHz.
This gives an excitation pulse that has a DC component, and spreads in spectrum to 5GHz @ -20dB. The graph shows voltage at port 1 (composite of incident + reflected) and the voltage at port 2. The line is well matched, so there is very little reflection in this example.

![pulse_line_dc](https://github.com/thliebig/openEMS-Project/assets/47922981/a3168bd2-3020-465e-82f9-ffe86eda427e)

>  What problems might be associated with fc being too wide?

Going too extreme, you might loose accuracy. Our FDTD simulation stops when the residual energy is below a threshold (often 1E-4 for -40dB). If we use an excitation where almost all energy is outside your band of interest, I would expect less accurate results for your band of interest because convergence check might stop too early. Not sure about that.

50 MHz excitation bandwidth for a 144 MHz yagi with 2 or 4 Mhz design bandwidth might be a bit extreme indeed, so I did a little experiment with different bandwidth for your yagi model. In the attached code, I tweaked the model a bit to run faster (switch in the code for coarse mesh) and then tested different excitation bandwidth:

BW 50 MHz took 2m:25s and 180k timesteps
BW 10 MHz took 3m:25s and 278k timesteps
BW  5 MHz took 4m:01s and 330k timesteps

The port voltages are plotted below for all 3 cases. You can see that the wider pulse envelope for smaller bandwidth takes longer time to settle below the threshold value where FDTD is considered converged.

fc=50 MHz:
![yagi_bw50_voltage](https://github.com/thliebig/openEMS-Project/assets/47922981/f563fafd-6b2b-4b4d-8af1-4c1e43e0404b)

fc=10 MHz:
![yagi_bw10_voltage](https://github.com/thliebig/openEMS-Project/assets/47922981/e6f60936-0beb-41c7-a3d4-b43c59358d49)

fc=5 MHz:
![yagi_bw5_voltage](https://github.com/thliebig/openEMS-Project/assets/47922981/2ea48a10-d732-41b8-b409-b59c275732f5)

S-parameters are not exactly the same, there is a little shift in resonance frequency. 

![compare_s11_bw50_bw10_bw5](https://github.com/thliebig/openEMS-Project/assets/47922981/abffa1cd-8675-4519-85f6-e1be0259c187)

One other change to previous models: this model uses PML (perfectly matched layers) on all sides because that seemed to make a difference to the simple MUR absorbing sheet. PML are better than MUR in absorbing incident waves from any angle, but they are also slower in simulation.
[yagi_LuboJ_modified_2_wire2mm_fastmesh_allPML_bw10.zip](https://github.com/thliebig/openEMS-Project/files/13728644/yagi_LuboJ_modified_2_wire2mm_fastmesh_allPML_bw10.zip)

**Reply by KJ7LNW** [reply]
Interesting.  So which resonant frequency do you expect to be the most accurate?  Would the 5MHz bandwidth be the most accurate even though it took longer since it is near the targeted bandwidth of 4MHz, or would wider bandwidth (like maybe 10MHz) be better, so the simulation accuracy doesn't fall off too quickly?

**Reply by VolkerMuehlhaus** [reply]
For most accurate results I would use 10 MHz bandwidth here, for the reason that you mentioned. 

But don't forget that in FDTD, we have many error sources like meshing and boundaries and ports and a few more. You know this from your MoM code, but here we discretize and solve the volume **around** the metal, up to some boundary, so using FDTD offers a wide range of possible error sources in the model.

To show another influence on resonance frequency, below is a comparison of
- "coarse mesh" model with 1 cell for the antenna element cross section and 15 cells/wavelength in air 
- "finer mesh" model with 2x2 cells for the antenna element cross section and 20 cells/wavelength in air 

This also results in a small frequency shift, and also in antenna patterns, but for practical use I would consider the S11 good enough to go ahead and build the antenna. 

![yagi_bw10_fine_coarse](https://github.com/thliebig/openEMS-Project/assets/47922981/2724d54b-22d3-4fc6-a3ed-cb22bb1eeb3a)

Regarding pattern, the effect of mesh is somewhat more pronounced, and might be relevant if you look at the backlobe. But I would not recommend to use a "global" fine mesh for the entire analysis volume, because that is inefficient and makes no difference in regions far away from the antenna elements. Instead I would add some local mesh refinement around the edges of the antenna elements, and leave the rest of the analysis volume at 20 cells/wavelength or less.

![pattern_coarse_fine](https://github.com/thliebig/openEMS-Project/assets/47922981/bd9845db-df5b-4c4d-8458-58b641ed94c9)

What I do for EM simulations in my job is some convergence analysis, where I check convergence of results using different mesh density, and then see what offset I get from using a coarse/moderate mesh density for a "good enough" working model. In that working model for tweaking/tuning/parameter sweep I then keep the offset to the very accurate model in mind. And only for the final model I then do another very fine mesh model for verification.

You will need to develop some experience for your yagi models, what model setting is critical and what is not. This can be really different between your "wire/thin tube" models and other antenna types.

Final note: please take my models with a grain of salt. In my job I am using FDTD for PCB-type antennas, and don't simulate yagis, but I did one hobby project for 432 yagi [a while ago](https://muehlhaus.com/support/ham/yagisimulation). That yagi simulation was done with another FDTD tool (not openEMS), so my experience with openEMS settings for yagi design is very limited, and here for openEMS I only extrapolate from my experience with that other FDTD tool ...

**Reply by VolkerMuehlhaus** [reply]
> But I would not recommend to use a "global" fine mesh for the entire analysis volume, because that is inefficient and makes no difference in regions far away from the antenna elements. Instead I would add some local mesh refinement around the edges of the antenna elements, and leave the rest of the analysis volume at 20 cells/wavelength or less.

Here is a model where I implemented these changes to increase the local mesh density near the elements, and keep moderate mesh density outside. The results for S11 obtained with that model are very close to the values that I get using my commercial FDTD tool with automatic meshing, resonance agrees to 0.2 MHz. The pattern agrees quite well to both the fine mesh pattern in my last post, and to the commercial FDTD tool results.

edit: Fixed mistake with calculation of the 2D pattern, there was an offset due to log10() missing
[yagi_wire2mm_fastmesh2_allPML_bw10.zip](https://github.com/thliebig/openEMS-Project/files/13745378/yagi_wire2mm_fastmesh2_allPML_bw10.zip)

![vswr_compare](https://github.com/thliebig/openEMS-Project/assets/47922981/b2c0f88e-f1a8-4b68-a3ad-62b1bae47aa3)

![pattern_compare](https://github.com/thliebig/openEMS-Project/assets/47922981/814f689c-e7d5-4e97-87b5-5e45186867f6)

I took away the "center" mesh lines that divided the 2mm x 2mm elements into 4 cells, and meshed the antenna element cross section  with just 1 cell. This has little effect on mesh count, but a large effect on FDTD step size because this was the smallest mesh cell in the entire model, and determines FDTD time step. By increasing that smallest mesh cell from 1mm to 2mm (=wire diameter), the simulation runs faster now, despite a larger total number of cells. 

Here is a view of the mesh strategy:

![extra_meshlines](https://github.com/thliebig/openEMS-Project/assets/47922981/298ba7d8-3c1e-4a5d-a87f-08573f5c1fc9)

![extra_meshlines_crosssection](https://github.com/thliebig/openEMS-Project/assets/47922981/51e350d4-df38-4002-95e0-ed6e3ddda573)

**Reply by KJ7LNW** [reply]
Awesome, thank you for the very well-developed and thoughtful process around the antenna design.  The comparison to a commercial product is encouraging as well!
